File: C++_Base_PT4/OOP2Struc7.cpp
#include "pt4.h"
#include <vector>

using namespace std;

class Component
{
public:
    virtual string Show() = 0;
};

// §²§¦§¡§­§ª§©§µ§«§´§¦ §¬§­§¡§³§³-§±§°§´§°§®§°§¬ ConcreteComponent
class ConcreteComponent: public  Component
{
public:
	ConcreteComponent(string s)
	{
		text = s;	
	}
	string Show(){
		return text;
	}	
private:
	string text;
};

class Decorator : public Component
{
protected:
    Component* comp;
};

class ConcreteDecoratorA : public Decorator
{
public:
	ConcreteDecoratorA(Component *comp)
	{
		this -> comp = comp;
	}
	string Show(){
		return "="+comp->Show()+"=";
	}
};

class ConcreteDecoratorB : public Decorator
{
public:
	ConcreteDecoratorB(Component *comp)
	{
		this -> comp = comp;
	}
	string Show(){
		return "("+comp->Show()+")";
	}
};

void Solve()
{
    Task("OOP2Struc7");
	int N;
	pt >> N;
	Component *arr[N];
	vector<string> v;
	for(int i = 0; i < N; ++i)
	{
		string S, D;
		pt >> S >> D;
		arr[i] = new ConcreteComponent(S);
		for(int j = 0; j < D.size(); j++)
		{
			if(D[j] == 'A')
			{
				ConcreteDecoratorA a = new ConcreteDecoratorA(arr[i]);	
				arr[i] = new ConcreteComponent(a.Show());
			}
			else{
				ConcreteDecoratorB b(arr[i]);				
				arr[i] = new ConcreteComponent(b.Show());
			}
		}
	}
	for(int i = N - 1; i >= 0; i--)
	{
		Show(arr[i]->Show());
		pt << arr[i]->Show();
	}
}

File: C#_Base_PT4/OOP2Struc7.cs
// File: "OOP2Struc7"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        public abstract class Component
        {
            public abstract string Show();
        }

        // Implement the ConcreteComponent descendant class
        public class ConcreteComponent : Component
        {
            string s = null;

            public ConcreteComponent(string s)
            {
                this.s = s;
            }
            public override string Show()
            {
                return s;
            }
        }

        public abstract class Decorator : Component
        {
            protected Component comp;
        }

        // Implement the ConcreteDecoratorA
        //   and ConcreteDecoratorB descendant classes
        public class ConcreteDecoratorA : Decorator
        {
            public ConcreteDecoratorA(Component comp)
            {
                this.comp = comp;
            }
            public override string Show()
            {
                //throw new NotImplementedException();
                return "==" + comp.Show() + "==";
            }
        } 

        public class ConcreteDecoratorB : Decorator
        {
            public ConcreteDecoratorB(Component comp)
            {
                this.comp = comp;
            }
            public override string Show()
            {
                //throw new NotImplementedException();
                return "(" + comp.Show() + ")";
            }
        } 
        public static void Solve()
        {
            Task("OOP2Struc7");
            int N = GetInt();
            Component[] arr = new Component[N];
            for(int i = 0; i < N; i++)
            {
                string S = GetString();
                string D = GetString();
                arr[i] = new ConcreteComponent(S);
                for(int j = 0; j < D.Length; j++)
                {
                    if(D[j] == 'A')
                    {
                        ConcreteDecoratorA a = new ConcreteDecoratorA(arr[i]);
                        arr[i] = new ConcreteComponent(a.Show());
                    }else{
                        ConcreteDecoratorB b = new ConcreteDecoratorB(arr[i]);
                        arr[i] = new ConcreteComponent(b.Show());
                    }
                }
            }

            for(int i = N-1; i >= 0; i--)
            {
                Put(arr[i].Show());
            }

        }
    }
}
